### PR TITLE
Refactor Ssn to implement IsoCodeInterface

### PR DESCRIPTION
**Pull request by enumag**
Alternative to https://github.com/ronanguilloux/IsoCodes/pull/34.

### CODE REVIEW

**Comment by Soullivaneuh** [comment]
Well, indeed it's even more simple.

But I think it's BC break because user won't be able to instantiate `Ssn` class without be warned.

**Comment by enumag** [comment]
@Soullivaneuh Nope, there is no warning. I believe it's not a BC break. https://3v4l.org/sP7Lk

**Comment by enumag** [comment]
We might want to consider using `static::` and `protected` instead of `self::` and `private` for extendability.

**Comment by Soullivaneuh** [comment]
> Nope, there is no warning. I believe it's not a BC break. https://3v4l.org/sFWY3

Maybe add a BC test for this would be great.

**Comment by enumag** [comment]
Done. There is something wrong with the Ssn tests though. They fail on my computer but it's not related to my patch. Thay fail on the current master branch as well.

**Comment by enumag** [comment]
The author should add .travis.yml and setup running tests by travis to prevent that in the future.

**Comment by Soullivaneuh** [comment]
I don't understand. Travis is set up and is actually green here: https://travis-ci.org/ronanguilloux/IsoCodes/builds/83485286

**Comment by Soullivaneuh** [comment]
Oh you mean deprecations error.

Yeah, I submitted PR https://github.com/ronanguilloux/IsoCodes/pull/57 for Travis improve.

I'll propose something better to manage deprecations after that. :+1:

**Comment by enumag** [comment]
No I don't mean deprecation errors. I mean actual asserts in SsnTest and CreditCardTest. Maybe it's windows-specific or something.

I'm sorry I missed the travis somehow.

**Comment by enumag** [comment]
fails

**Comment by Soullivaneuh** [comment]
Hmm, that's strange.

Unfortunately, we can't reproduce it on Travis, Windows is not available.

**Comment by enumag** [comment]
As for SsnTest the errors disappear if I switch the Ssn.php file from CRLF to LF. That should be considered a bug though...

Dunno what's the problem with CreditCard.

**Comment by Soullivaneuh** [comment]
> the errors disappear if I switch the Ssn.php file from CRLF to LF.

Could you please open a separate issue with more details about this?

**Comment by enumag** [comment]
I've opened #74 and #75.

**Comment by enumag** [comment]
@Soullivaneuh Anyway since the failing tests are not related let's get back to the topic. Do you consider this solution to be better than #34? If not, why? In my opinion we should decide which is better and close the other one.

**Comment by Soullivaneuh** [comment]
Your solution is more simple and cleaner IMO.

But, even if `Ssn` class could be instantiated with that, there is no deprecation warning thrown about that.

This is my objective opinion. I let @ronanguilloux get the last decision. :-)

**Comment by enumag** [comment]
Well adding a deprecated error to __construct is trivial. ;-)

**Comment by Soullivaneuh** [comment]
> Well adding a deprecation warning to __construct is trivial. ;-)

Indeed. Let's do it so! :-)

**Comment by enumag** [comment]
@Soullivaneuh Done.

File: src/IsoCodes/Ssn.php [hunk 1]
```diff
@@ -211,17 +211,22 @@ class Ssn
...
     // The SSA uses a funky method of figuring out what group number to use next. This area has them in the proper order and makes it easier to generate a SSN.
-    public $possibleGroups = array(1, 3, 5, 7, 9, 10, 12, 14, 16, 18, 20, 22, 24, 26, 28, 30, 32, 34, 36, 38, 40, 42, 44, 46, 48, 50, 52, 54, 56, 58, 60, 62, 64, 66, 68, 70, 72, 74, 76, 78, 80, 82, 84, 86, 88, 90, 92, 94, 96, 98, 2, 4, 6, 8, 11, 13, 15, 17, 19, 21, 23, 25, 27, 29, 31, 33, 35, 37, 39, 41, 43, 45, 47, 49, 51, 53, 55, 57, 59, 61, 63, 65, 67, 69, 71, 73, 75, 77, 79, 81, 83, 85, 87, 89, 91, 93, 95, 97, 99);
+    public static $possibleGroups = array(1, 3, 5, 7, 9, 10, 12, 14, 16, 18, 20, 22, 24, 26, 28, 30, 32, 34, 36, 38, 40, 42, 44, 46, 48, 50, 52, 54, 56, 58, 60, 62, 64, 66, 68, 70, 72, 74, 76, 78, 80, 82, 84, 86, 88, 90, 92, 94, 96, 98, 2, 4, 6, 8, 11, 13, 15, 17, 19, 21, 23, 25, 27, 29, 31, 33, 35, 37, 39, 41, 43, 45, 47, 49, 51, 53, 55, 57, 59, 61, 63, 65, 67, 69, 71, 73, 75, 77, 79, 81, 83, 85, 87, 89, 91, 93, 95, 97, 99);
+
+    public function __construct()
+    {
+        trigger_error('Instantiating Ssn validator is deprecated since version 1.3. Use Ssn::validate($ssn) instead.', E_USER_DEPRECATED);
```
Comment (Soullivaneuh) [inline src/IsoCodes/Ssn.php]: * Deprecated error should be silenced (`@`).
* Use FQN for `Ssn` class name

Comment (enumag) [inline src/IsoCodes/Ssn.php]: I was using the existing deprecated errors for reference - https://github.com/ronanguilloux/IsoCodes/blob/master/src/IsoCodes/Isbn10.php. There is no `@` and FQN is not used. You can put such changes in separate PR if you wish.

Comment (Soullivaneuh) [inline src/IsoCodes/Ssn.php]: Yeah. This was made before [phpunit-bridge](https://github.com/symfony/phpunit-bridge) requires deprecated errors to be silenced.

I'll do a PR but please correct it on this one.

Ok for non-FQN.

Comment (enumag) [inline src/IsoCodes/Ssn.php]: You mean to add a `@` to the test right? Putting it here would suppress the warning for users.

**Comment by Soullivaneuh** [comment]
Coverage decreased. Could you please take a look?

**Comment by enumag** [comment]
I think that's because of the non-silenced deprecated errors. There is no way for this commit to really decrease code coverage.

**Comment by Soullivaneuh** [comment]
@ronanguilloux https://coveralls.io/builds/3761491/source?filename=IsoCodes%2FSsn.php

> The file "IsoCodes/Ssn.php" isn't available on github. Either it's been removed, or the repo root directory needs to be updated.

Could you fix that? Just have to put `src` as root path on coveralls configuration.
